### PR TITLE
LPD-17343 Performance fix

### PR DESCRIPTION
**Pull request by pei-jung**
/cc @tinatian @victorhcunha

### CODE REVIEW

**Comment by pei-jung** [comment]
ci:test:relevant

**Comment by pei-jung** [comment]
ci:test:publications

File: modules/apps/dispatch/dispatch-web/src/main/java/com/liferay/dispatch/web/internal/display/context/DispatchTriggerDisplayContext.java [hunk 1]
```diff
@@ -225,11 +229,24 @@ public SearchContainer<DispatchTrigger> getSearchContainer() {
...
 		_searchContainer.setOrderByType(getOrderByType());
 		_searchContainer.setResultsAndTotal(
-			() -> _dispatchTriggerLocalService.getDispatchTriggers(
-				dispatchRequestHelper.getCompanyId(),
-				_searchContainer.getStart(), _searchContainer.getEnd()),
-			_dispatchTriggerLocalService.getDispatchTriggersCount(
-				dispatchRequestHelper.getCompanyId()));
+			ListUtil.filter(
+				_dispatchTriggerLocalService.getDispatchTriggers(
+					dispatchRequestHelper.getCompanyId(), QueryUtil.ALL_POS,
```
Comment (tinatian) [inline modules/apps/dispatch/dispatch-web/src/main/java/com/liferay/dispatch/web/internal/display/context/DispatchTriggerDisplayContext.java:234]: @pei-jung `QueryUtil.ALL_POS` is wrong, please make it consistent with before by using `_searchContainer.getStart(), _searchContainer.getEnd()`

Comment (pei-jung) [reply modules/apps/dispatch/dispatch-web/src/main/java/com/liferay/dispatch/web/internal/display/context/DispatchTriggerDisplayContext.java:234]: This is how [searchContainer.setResultsAndTotal(List results)](https://github.com/liferay/liferay-portal/blob/master/portal-kernel/src/com/liferay/portal/kernel/dao/search/SearchContainer.java#L488-L492) works. It expects the entire results set in order to get the total count. The results list will be paginated with the start and end values.

Comment (tinatian) [reply modules/apps/dispatch/dispatch-web/src/main/java/com/liferay/dispatch/web/internal/display/context/DispatchTriggerDisplayContext.java:234]: @pei-jung So it means "_searchContainer.getStart(), _searchContainer.getEnd()" is wrong? I am confused here, which part of logic make it consistent with "_searchContainer.getStart(), _searchContainer.getEnd()"?

Comment (pei-jung) [reply modules/apps/dispatch/dispatch-web/src/main/java/com/liferay/dispatch/web/internal/display/context/DispatchTriggerDisplayContext.java:234]: @tinatian the previous code calls `setResultsAndTotal(UnsafeSupplier<List<R>, E> setResultsUnsafeSupplier, int total)`, which sets the `results` and `totoal` fields separately. It calls `_dispatchTriggerLocalService.getDispatchTriggers(dispatchRequestHelper.getCompanyId(), _searchContainer.getStart(), _searchContainer.getEnd())` to get the paginated list for the table display and `_dispatchTriggerLocalService.getDispatchTriggersCount` for the total count. 

However, in order to apply the feature flag filter, we need to get all the dispatch triggers first, filter them by feature flag, then paginate it. We'll be calling the same service for the paginated list and total count, so I used `searchContainer.setResultsAndTotal(List results)` to save a service call. It still uses `_searchContainer.getStart(), _searchContainer.getEnd()` internally to set the paginated list. 

```
_setResults(results.subList(_start, _resultEnd));
```

Comment (tinatian) [reply modules/apps/dispatch/dispatch-web/src/main/java/com/liferay/dispatch/web/internal/display/context/DispatchTriggerDisplayContext.java:234]: @pei-jung Thanks. I see... the API method name is so confusing here, you have to go down into its impl to tell the difference.

**Comment by tinatian** [comment]
@pei-jung Can you verify once again if the startup regression caused by this pull is fixed? Once you confirmed that, this pull is good to go. 

Thanks.
Tina

**Comment by JerryAdrianoo** [comment]
**Liferay Portal CE Version - Before** 
| Open count of Prepared Statement | count of Prepared Statement Query |
|--------|--------|
| 64 | 96 |

------------------------------------------------------

 **Liferay Portal CE Version - Now LPD-17343** 
| Open count of Prepared Statement | count of Prepared Statement Query |
|--------|--------|
| 64 | 96 |

Was solved. the previous numbers with regression were 70 and 105 respectively.
@tinatian Do you want any more information about it? if it's not enough, please let me know.

**Comment by tinatian** [comment]
@JerryAdrianoo The test you did is wrong. 
What you need to do: 
1. Update the latest upstream, do a profile
2. Check out this pull on latest upstream, do a profile
Show me the diff of the two profiles.

**Comment by JerryAdrianoo** [comment]
> @JerryAdrianoo The test you did is wrong. 
> What you need to do: 
> 1. Update the latest upstream, do a profile
> 2. Check out this pull on latest upstream, do a profile
> Show me the diff of the two profiles.
> 

Ok the second step was what I did above. I'll do step 1 and show the difference.

**Comment by JerryAdrianoo** [comment]
**Liferay Portal CE Latest Upstream** 
| Open count of Prepared Statement | count of Prepared Statement Query |
|--------|--------|
| 64 | 96 |

------------------------------------------------------

 **Liferay Portal CE - Now with these changes** 
| Open count of Prepared Statement | count of Prepared Statement Query |
|--------|--------|
| 64 | 96 |

@tinatian No diff in the other fields in the profile target was found. _Open Count of SQL, Open File Number_, and _Thread
Create Count_ is stable. Is this correct now?
[Snapshots](https://drive.google.com/drive/folders/1x-LxCcGW1c1t_y8qkoLha3jDovULTZFE?usp=sharing)

**Comment by tinatian** [comment]
@JerryAdrianoo Yes, this is what I want and it is enough. 

Thanks.
Tina

**Comment by tinatian** [comment]
See https://github.com/brianchandotcom/liferay-portal/pull/148415

**Comment by pei-jung** [comment]
Thanks @tinatian @JerryAdrianoo!

**Pull request closed without merging**